### PR TITLE
[Aptos Framework] [Account] emit events upon key rotation

### PR DESCRIPTION
**Pull request by chloeqjz**
### Test Plan

- - -
This change is [Reviewable](https://reviewable.io/reviews/aptos-labs/aptos-core/3408)

### CODE REVIEW

**Review by davidiw — COMMENTED** [review]
this is going to be one big ugly change... you need to modify 
https://github.com/aptos-labs/aptos-core/blob/main/types/src/account_config/resources/core_account.rs

you also need to regenerate goldens in
api, aptos-move/, e2e-testsuite, ecosystem/sf-indexer
UP=1 in aptos-transactional-test-harness

File: aptos-move/framework/aptos-framework/sources/reconfiguration.move [hunk 1]
```diff
@@ -53,7 +53,7 @@ module aptos_framework::reconfiguration {
         system_addresses::assert_aptos_framework(aptos_framework);
 
         // assert it matches `new_epoch_event_key()`, otherwise the event can't be recognized
-        assert!(guid::get_next_creation_num(signer::address_of(aptos_framework)) == 1, error::invalid_state(EINVALID_GUID_FOR_EVENT));
+        assert!(guid::get_next_creation_num(signer::address_of(aptos_framework)) == 2, error::invalid_state(EINVALID_GUID_FOR_EVENT));
```
Comment (areshand) [inline aptos-move/framework/aptos-framework/sources/reconfiguration.move:56]: why do we need to change to 2?  new_epoch_event_key is hard coded at https://github.com/aptos-labs/aptos-core/blob/5f3ac3920ce7fb9865f8357f04fe8c2fd93af769/types/src/on_chain_config/mod.rs#L158 and have many dependencies across codebase.

**Comment by chloeqjz** [comment]
> this is going to be one big ugly change... you need to modify https://github.com/aptos-labs/aptos-core/blob/main/types/src/account_config/resources/core_account.rs
> 
> you also need to regenerate goldens in api, aptos-move/, e2e-testsuite, ecosystem/sf-indexer UP=1 in aptos-transactional-test-harness

wait what does UP=1 mean?

File: aptos-move/framework/aptos-framework/sources/reconfiguration.move (hunk 1 above)
Comment (chloeqjz) [reply aptos-move/framework/aptos-framework/sources/reconfiguration.move:56]: changing all the other hard coded values too! I think it's bc we are adding a new event in account and it's created first, so we need to add 1 to all the other event keys

File: aptos-move/framework/aptos-framework/sources/account.move [hunk 2]
```diff
@@ -261,6 +269,19 @@ module aptos_framework::account {
...
         account_resource.authentication_key = new_auth_key;
+
+        // Emit event upon key rotation
+        let account = borrow_global_mut<Account>(addr);
+        event::emit_event<KeyRotationEvent>(
+            &mut account.key_rotation_events,
+            KeyRotationEvent {
+                // Address of rotator may be different from the addr of the account
+                // once we allow key rotation on behalf of another account
+                address_of_rotator: addr,
```
Comment (areshand) [inline aptos-move/framework/aptos-framework/sources/account.move:280]: address_of_rotator can be found from ledger info which is the same as `account: &signer`, no need to store it.

File: aptos-move/framework/aptos-framework/sources/account.move [hunk 3]
```diff
@@ -69,6 +70,12 @@ module aptos_framework::account {
         recipient_address: address,
     }
 
+    struct KeyRotationEvent has drop, store {
+        address_of_rotator: address,
```
Comment (lightmark) [inline aptos-move/framework/aptos-framework/sources/account.move:74]: name as `origin`?

**Pull request closed without merging**